### PR TITLE
feat: Save alerts from 3rd party providers

### PR DESCRIPTION
**Pull request by Matvey-Kuk**
Saving alerts from 3rd party providers, relying on duplicator logics.

It's a PR to another PR https://github.com/keephq/keep/pull/1276

### CODE REVIEW

**Comment by Matvey-Kuk** [comment]
Test `test_null_handling` is failing in the upstream

File: keep/api/routes/preset.py [hunk 1]
```diff
@@ -60,6 +64,51 @@
     ),
 }
 
+async def pull_alerts_from_providers(
+    tenant_id: str, pusher_client: Pusher | None, sync: bool = False
```
Comment (talboren) [inline keep/api/routes/preset.py:68]: redundant, isn't it?
```suggestion
    tenant_id: str
```

Comment (talboren) [reply keep/api/routes/preset.py:68]: or maybe send a "poll-alerts" (or is it happening afterwards?)

**Comment by shahargl** [comment]
@Matvey-Kuk ill review the test it’s on me

File: keep/api/routes/preset.py [hunk 2]
```diff
@@ -188,10 +237,19 @@ def update_preset(
...
+    bg_tasks: BackgroundTasks,
     preset_name: str,
     authenticated_entity: AuthenticatedEntity = Depends(AuthVerifier()),
 ) -> list[AlertDto]:
+    
+    # Gathering alerts may take a while and we don't care if it will finish before we return the response.
+    # In the worst case, gathered alerts will be pulled in the next request.
+    bg_tasks.add_task(
+        pull_alerts_from_providers,
+        authenticated_entity.tenant_id, None, sync=True
```
Comment (talboren) [inline keep/api/routes/preset.py:250]:
```suggestion
        authenticated_entity.tenant_id
```

File: keep/api/routes/preset.py (hunk 1 above)
Comment (talboren) [reply keep/api/routes/preset.py:68]: looks like it just uses process_event which does the work :)

File: keep/api/routes/preset.py [hunk 3]
```diff
@@ -1,16 +1,20 @@
 import logging
 
-from fastapi import APIRouter, Depends, HTTPException
+from pusher import Pusher
```
Comment (talboren) [inline keep/api/routes/preset.py:3]:
```suggestion
```

File: keep/api/routes/preset.py [hunk 4]
```diff
@@ -60,6 +64,51 @@
...
+                {},
+                tenant_id,
+                None,
+                None,
+                fingerprint,
+                None,
+                None,
+                alert,
+                save_if_duplicate=False,
+            )
```
Comment (talboren) [inline keep/api/routes/preset.py:110]: Another thing here (don't know how important it is but..) is process_event accepts a list of AlertDto, so you can:
```suggestion
        alerts = [alert for alert in sorted_provider_alerts_by_fingerprint.values()]
        await process_event(
            {},
            tenant_id,
            None,
            None,
            fingerprint,
            None,
            None,
            alerts,
            save_if_duplicate=False,
        )
```
or something, but meh :P

**Review by talboren — APPROVED** [review]
left a few small comment, other than that, LGTM!

**Comment by shahargl** [comment]
tests were fixed - https://github.com/keephq/keep/pull/1281

File: keep/api/routes/preset.py (hunk 1 above)
Comment (Matvey-Kuk) [reply keep/api/routes/preset.py:68]: <3

File: keep/api/routes/preset.py (hunk 4 above)
Comment (Matvey-Kuk) [reply keep/api/routes/preset.py:110]: I didn't want to re-calculate fingerprint there, not sure if it's critical or not 🤔

**Pull request merged**